ml-algorithms/src/main/java/org/opensearch/ml/engine/encryptor/EncryptorImpl.java

```diff
@@ -225,4 +166,264 @@ private void initMasterKey(String tenantId) {
+            .tenantId(tenantId)
+            .index(ML_CONFIG_INDEX)
+            .id(masterKeyId)
+            .overwriteIfExists(false)
+            .dataObject(
+                Map
+                    .of(
+                        MASTER_KEY,
+                        generatedMasterKey,
+                        CREATE_TIME_FIELD,
+                        Instant.now().toEpochMilli(),
+                        TENANT_ID_FIELD,
+                        Objects.requireNonNullElse(tenantId, DEFAULT_TENANT_ID)
+                    )
+            )
+            .build();
+    }
+
+    private void handlePutDataObjectResponse(
+        String tenantId,
+        String masterKeyId,
+        ThreadContext.StoredContext context,
+        PutDataObjectResponse putDataObjectResponse,
+        Throwable throwable,
+        AtomicReference<Exception> exceptionRef,
+        CountDownLatch latch,
+        String generatedMasterKey
+    ) throws IOException {
+        context.restore();
+
+        if (throwable != null) {
+            handlePutDataObjectFailure(tenantId, masterKeyId, context, throwable, exceptionRef, latch);
+        } else {
+            IndexResponse indexResponse = IndexResponse.fromXContent(putDataObjectResponse.parser());
+            log.info("Master key creation result: {}, Master key id: {}", indexResponse.getResult(), indexResponse.getId());
+            this.tenantMasterKeys.put(Objects.requireNonNullElse(tenantId, DEFAULT_TENANT_ID), generatedMasterKey);
+            log.info("ML encryption master key initialized successfully");
+            latch.countDown();
+        }
+    }
+
+    private void handlePutDataObjectFailure(
+        String tenantId,
+        String masterKeyId,
+        ThreadContext.StoredContext context,
+        Throwable throwable,
+        AtomicReference<Exception> exceptionRef,
+        CountDownLatch latch
+    ) {
+        Exception cause = SdkClientUtils.unwrapAndConvertToException(throwable, OpenSearchStatusException.class);
+        if (cause instanceof VersionConflictEngineException) {
+            handleVersionConflict(tenantId, masterKeyId, context, exceptionRef, latch);
+        } else {
+            log.debug("Failed to index ML encryption master key to config index", cause);
+            exceptionRef.set(cause);
+            latch.countDown();
+        }
+    }
+
+    private void handleVersionConflict(
+        String tenantId,
+        String masterKeyId,
+        ThreadContext.StoredContext context,
+        AtomicReference<Exception> exceptionRef,
+        CountDownLatch latch
+    ) {
+        sdkClient
+            .getDataObjectAsync(
+                createGetDataObjectRequest(tenantId, new FetchSourceContext(true, Strings.EMPTY_ARRAY, Strings.EMPTY_ARRAY))
+            )
+            .whenComplete((response, throwable) -> {
+                try {
+                    handleVersionConflictResponse(tenantId, masterKeyId, context, response, throwable, exceptionRef, latch);
```

Comment:
one question when handling version conflict. We call **getDataObjectAsync(getDataObjectRequest)** -> and if empty -> **initializeNewMasterKey** -> **putDataObjectAsync** ->  find **VersionConflictEngineException** -> **handleVersionConflict** -> **getDataObjectAsync** -> **handleVersionConflictResponse**, does the two **getDataObjectAsync** totally same? If so, we will also get empty and looks we cannot handle version conflict. Do I miss something?